proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,203 @@
+(C, labels { result [t2*], kind catch } ⊢ instr''* : []→[t2*])?
+-----------------------------------------------------------------------------
+C ⊢ try bt instr* (catch x instr'*)* (catch_all instr''*)? end : [t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels {result [t2*], kind try} ⊢ instr* : [t1*]→[t2*]
+C.labels[l].kind = try
+----------------------------------------------------------
+C ⊢ try bt instr* delegate l : [t1*]→[t2*]
+```
+
+## Execution (Reduction)
+
+### Runtime structure
+
+Stores
+
+```
+S ::= {..., tags taginst*}
+```
+
+Tag Instances
+
+```
+taginst ::= {type tagtype}
+```
+
+Module Instances
+
+```
+m ::= {..., tags a*}
+```
+
+Administrative Instructions
+
+```
+instr ::= ... | throw a | catch_n{a? instr*}* instr* end
+        | delegate{l} instr* end | caught_m{a val^n} instr* end
+```
+
+Block contexts and label kinds
+
+So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts, mirroring the label kinds of labels in validation contexts.
+
+```
+B^k ::= val* B'^k instr*
+B'^0 ::= [_]
+B'^{k+1} ::= label_n{instr*} B^k end | catch_m{a? instr*}* B^{k+1} end | caught_m{a val*} B^{k+1} end | delegate{l} B^{k+1} end
+```
+
+Note that `label_n{instr*} label_kind? [_] end? end` could be seen as a simplified control frame.
+
+(Alternatively, we could have the above `label_kind`s be also labels,  remove the additional `label_m` from the execution rules below, and remove the execution rules below where the new administrative instructions only contain `val*`. This would make labels even more similar to control frames.)
+
+Throw Contexts
+
+```
+T ::= val* [_] instr* | label_n{instr*} T end | caught_m{a val^n} T end
+   | frame_n{F} T end
+```
+
+### Instructions
+
+
+```
+F; throw x  ↪  F; throw a  (if F.module.tagaddrs[x]=a)
+
+caught_m{a val^n} B^l[rethrow l] end
+  ↪ caught_m{a val^n} B^l[val^n (throw a)] end
+
+caught_m{a val^n} val^m end  ↪  val^m
+```
+
+An absent tagaddr in a `catch_m` clause (i.e., `a? = ε`) represents a `catch_all`.
+
+```
+F; val^n (try bt instr* (catch x instr'*)* (catch_all instr''*)? end)
+  ↪  F; label_m{} (catch_m{a instr'*}*{instr''*}? val^n instr* end) end
+  (if bt = [t1^n]→[t2^m] ∧ (F.module.tagaddrs[x]=a)*)
+
+catch_m{a? instr*}* val^m end ↪ val^m
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; caught_m{a val^n} (val^n)? instr* end
+  (if (a1? = ε ∨ a1? = a) ∧ S.tags(a).type = [t^n]→[])
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; catch_m{a'? instr'*}* T[val^n (throw a)] end
+  (if a1? ≠ ε ∧ a1? ≠ a)
+
+S; F; catch_m T[val^n (throw a)] end
+  ↪  S; F; val^n (throw a)
+
+
+val^n (try bt instr* delegate l)
+  ↪ label_m{} (delegate{l+1} val^n instr* end) end
```

Comment:
Why `l+1`? It would help if we have a brief description on what the administrative instruction `delegate{}` is (not the Wasm instruction `delegate`). The same for the other administrative instructions too

Reply:
> Why `l+1`?

The `l` becomes `l+1` because we are adding a label around it, so we can break to it from inside `instr*`. 

> It would help if we have a brief description on what the administrative instruction delegate{} is (not the Wasm instruction delegate). The same for the other administrative instructions too

The administrative instructions `catch_n ... end`, `caught_n ... end` and `delegate ... end` are label-like, in that they (plus a surrounding label), delimit an instruction sequence, while also adding information on the special behaviour of these label - label-like combinations.

Off the top of my head perhaps we it's possible to merge each of these with its label, and instead define new labels. So in particular redefine labels to be:

```
label ::= label_n{instr*} |  catch_n{a^? instr*}* |  delegate_n{l} | caught_n{a val*}
```

WDYT?
(cc. @rossberg)

Reply:
Yeah, this is another indication that the hack I suggested for typing catch labels – which is to put the block label outside the catch – is not the greatest idea (see also [this comment]).

I'd much prefer to keep the label statement separate. If you treated all of these as labels, all of them would need to carry a continuation. And likewise future block-constructs we might add. That conflates concerns, increasing complexity.

Unfortunately, there is some conflation due to `rethrow` addressing its source `catch` via a label. (If we hadn't done that, i.e., only count catch clauses, we wouldn't have this problem.) That's why we need the ad-hoc annotation on the label in the context, and that's where the reduction hack comes from.

I'll try to come up with some replacement for this hack – dealing with a typing corner case shouldn't bleed into reduction rules in counter-intuitive ways.

Reply:
@aheejin, @rossberg, thinking about it, is the +1 really necessary? If we look at `delegate` as:

1. catching an `exn`,
2. "breaking" _inside_ the `l`th surrounding block,
3. and throwing `exn` from there, 

then I think we don't need the `+1`. In fact I think the reduction of administrative delegate is wrong to throw outside of the `l+1`'th surrounding block. WDYT?

Reply:
Hm, you may be right that there is an off-by-1 bug that could work in our favour. Should be easy to check by working through an example. And of course with the interpreter, provided it matches the rules faithfully.

Reply:
Apologies for the long delay!

Please never mind my previous explanation. I looked at this again and realised that the reduction rule of the administrative `delegate{ l }` should resemble the reduction of `br l`:

```
label_m{} B^l[ delegate{l} T[val^n (throw a)] end ] end
  ↪ val^n (throw a)
```

This way, there is a unique `B^l` that matches the reduction rule.

I modified Example 4 and added Example 5 in `Exceptions-formal-examples.md` which show the reduction steps of the administrative `delegate`s, to check that this works, and modified (or rather simplified) the interpreter accordingly, adding a couple of extra tests there.

I think that now it matches the idea of break-and-throw better. Note that the interpreter part that was removed didn't previously have a corresponding reduction rule in this formal-overview.